src/Transform/ONNX/Recompose.cpp

```diff
@@ -0,0 +1,392 @@
+//
+// TODO: This file is quite busy as the number of decomposing op is increasing.
+// It is better to move decomposition of each operation into a separate file.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Transform/ONNX/Recompose.hpp"
+#include "src/Pass/Passes.hpp"
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/Support/Debug.h"
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include "src/Support/TypeUtilities.hpp"
+
+#define DEBUG_TYPE "recompose"
+
+using namespace mlir;
+
+namespace {
+/// Include the patterns defined in the Declarative Rewrite framework.
+// #include "src/Transform/ONNX/ONNXRecompose.inc"
+
+// Support for recognizing patterns. Detects if the operation "op" has an input
+// operand number "matchThisOperandIndex" that is defined by an operation of
+// type "OP". If that is the case, "matchOperand" will be set to that operand,
+// and "matchOp" will be set to that op. For unary operations; write matches
+// only on success.
+//
+// This call is formatted so that it mimic the operations that we are trying to
+// match. For example:
+//
+// %norm = "onnx.Div"(%d, %stdDev)
+// %normScaled = "onnx.Mul"(%norm, %scale)
+//
+// We can have a match like this (version below is for binary op):
+//
+//  if (!operandOfOpDefinedBy<ONNXDivOp>(mulOp, norm, scale, divOp, 0))
+//
+// Namely, test if the mul op has input operands 0 that is defined by a divide
+// op. If it does, then set norm, scale, and divOp to their appropriate values.
+
+template <typename OP>
+bool operandOfOpDefinedBy(Operation *op, Value &matchOperand,
+    Operation *&matchOp, int64_t matchThisOperandIndex = 0) {
+  assert(matchThisOperandIndex >= 0 &&
+         matchThisOperandIndex < op->getNumOperands() &&
+         "bad match operand index");
+  Value operand = op->getOperand(matchThisOperandIndex);
+  // operand.dump();
+  //  Check for a match with definition of operand.
+  if (!operand.isa<BlockArgument>() && isa<OP>(operand.getDefiningOp())) {
+    matchOperand = operand;
+    matchOp = operand.getDefiningOp();
+    return true;
+  }
+  return false;
+}
+
+// Similar as above, for binary operation; write matches only on success.
+template <typename OP>
+bool operandOfOpDefinedBy(Operation *op, Value &matchOperand0,
+    Value &matchOperand1, Operation *&matchOp, int64_t matchThisOperandIndex) {
+  Value dummy;
+  if (operandOfOpDefinedBy<OP>(op, dummy, matchOp, matchThisOperandIndex)) {
+    matchOperand0 = op->getOperand(0);
+    matchOperand1 = op->getOperand(1);
+    return true;
+  }
+  return false;
+}
+
+struct RecomposeLayerNormFromAddPattern : public OpRewritePattern<ONNXAddOp> {
```

Comment:
Do you think this rewrite patten can be moved to Rewrite.cpp? It's like fusing Add into LayerNorm.

Reply:
good suggestion.

Reply:
done